torchio/transforms/preprocessing/spatial/cropOrPad.py

```diff
@@ -0,0 +1,151 @@
+        :return : 6 coordinates of a 3D bounding box (rmin, rmax, cmin, xmax, zmin, zmax)
+        """
+        r = np.any(mask_volume, axis=(1, 2))
+        c = np.any(mask_volume, axis=(0, 2))
+        z = np.any(mask_volume, axis=(0, 1))
+
+        rmin, rmax = np.where(r)[0][[0, -1]]
+        cmin, cmax = np.where(c)[0][[0, -1]]
+        zmin, zmax = np.where(z)[0][[0, -1]]
+
+        return rmin, rmax, cmin, cmax, zmin, zmax
+
+    @staticmethod
+    def _get_sample_shape(sample: dict) -> Tuple[int]:
+        """Return the shape of the first image in the sample."""
+        check_consistent_shape(sample)
+        for image_dict in sample.values():
+            if not is_image_dict(image_dict):
+                continue
+            data = image_dict[DATA].shape[1:]  # remove channels dimension
+            break
+        return data
+
+    @staticmethod
+    def _get_six_bounds_parameters(parameters: np.ndarray):
+        r"""Compute bounds parameters for ITK filters.
+
+        Args:
+            parameters: Tuple :math:`(d, h, w)` with the number of voxels to be
+                cropped or padded.
+
+        Returns:
+            Tuple :math:`(d_{ini}, d_{fin}, h_{ini}, h_{fin}, w_{ini}, w_{fin})`,
+            where :math:`n_{ini} = \left \lceil \frac{n}{2} \right \rceil` and
+            :math:`n_{fin} = \left \lfloor \frac{n}{2} \right \rfloor`.
+
+        Example:
+            >>> p = np.array((4, 0, 7))
+            >>> _get_six_bounds_parameters(p)
+            (2, 2, 0, 0, 4, 3)
+
+        """
+        parameters = parameters / 2
+        result = []
+        for n in parameters:
+            ini, fin = int(np.ceil(n)), int(np.floor(n))
+            result.extend([ini, fin])
+        return tuple(result)
+
+    def _compute_center_crop_or_pad(self, sample: dict):
+        source_shape = self._get_sample_shape(sample)
+        # The parent class turns the 3-element shape tuple (d, h, w)
+        # into a 6-element bounds tuple (d, d, h, h, w, w)
+        target_shape = np.array(self.bounds_parameters[::2])
+        diff_shape = target_shape - source_shape
+
+        cropping = -np.minimum(diff_shape, 0)
+        if cropping.any():
+            cropping_params = self._get_six_bounds_parameters(cropping)
+
+        padding = np.maximum(diff_shape, 0)
+        if padding.any():
+            padding_params = self._get_six_bounds_parameters(padding)
+
+        return padding_params, cropping_params
+
+    def _compute_mask_center_crop_or_pad(self, sample: dict):
+        mask = sample[self.mask_key][DATA].numpy()
+        # Original sample shape (from mask shape)
+        sample_shape = np.squeeze(mask).shape
+        # Calculate the bounding box of the brain
+        xmin, xmax, ymin, ymax, zmin, zmax = self._bbox_mask(np.squeeze(mask))
+        # Coordinates of the center of the brain
+        center_mask = ((xmax - xmin) / 2 + xmin, (ymax - ymin) / 2 + ymin, (zmax - zmin) / 2 + zmin)
```

Comment:
Can this be vectorized with NumPy?

Reply:
Which part do you wish to vectorize ? The calculation of the center given a min and max ?

Reply:
What about something like

```python
# Make self._bbox_mask return two NumPy arrays
bb_min, bb_max = self._bbox_mask(mask[0])  # squeeze might remove other dimensions than the first
center_mask = (bb_max - bb_min) / 2 + bb_min
```

Reply:
Ok, it is done.